Extend/Model/PaymentGateway.php

```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
-        $oTransaction->setRedirect($oRedirect);
+        $oTransaction->setRedirect(self::getRedirectUrls($oSession, $sShopUrl, $sSid));
 
         $oShopconfig = $this->getConfig();
         $oCurrency = $oShopconfig->getActShopCurrencyObject();
-        $oTransaction->setAmount(new Amount($fAmount, $oCurrency->name));
-
-        $oBasket = $oSession->getBasket();
+        $oTransaction->setAmount(new Amount($oBasket->getPrice()->getBruttoPrice(), $oCurrency->name));
 
         $sOrderDetails = $oOrder->oxorder__oxremark->value;
+
         if (!empty($sOrderDetails)) {
             $oTransaction->setOrderDetail($sOrderDetails);
         }
 
-        $oPayment = oxNew(Payment::class);
-        $oPayment->load($sPaymentMethod);
-
         if ($oPayment->oxpayments__wdoxidee_additional_info->value) {
-            $this->_addAdditionalInfo($oTransaction, $oOrder, $oPayment);
+            self::_addAdditionalInfo($oTransaction, $oOrder, $oPayment, $oSession->getId());
         }
 
         if ($oPayment->oxpayments__wdoxidee_descriptor->value) {
-            $sDescriptor = self::getDescriptor($oOrder->oxorder__oxid->value);
-            $oTransaction->setDescriptor($sDescriptor);
+            self::_addDescriptor($oTransaction, $oOrder->oxorder__oxid->value);
         }
 
         if ($oPayment->oxpayments__wdoxidee_basket->value) {
-            $this->_addBasketInfo($oTransaction, $oBasket);
+            self::_addBasketInfo($oTransaction, $oBasket);
         }
 
         $oTransaction->setNotificationUrl($sShopUrl
             . 'index.php?cl=wcpg_notifyhandler&fnc=handleRequest&pmt='
             . Payment_Method::getOxidFromSDKName($oPaymentMethod->getTransaction()->getConfigKey()));
 
+        return $oTransaction;
+    }
+
+    /**
+     * Calls the {@link TransactionService} and processes the {@link Transaction}
+     *
+     * @param Transaction $oTransaction
+     * @param Order       $oOrder
+     *
+     * @return FailureResponse|FormInteractionResponse|InteractionResponse|Response|SuccessResponse
+     * @throws Exception
+     */
+    public function executeTransaction($oTransaction, $oOrder)
+    {
+        $oSession = $this->getSession();
+        $oBasket = $oSession->getBasket();
+        /*
+         * @var $oPayment Payment
+         */
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($oBasket->getPaymentId());
+
+        $oPaymentMethod = Payment_Method_Factory::create($oPayment->oxpayments__oxid->value);
+
+        $oTransactionConfig = $oPaymentMethod->getConfig($oPayment);
+        $oTransactionService = new TransactionService($oTransactionConfig, $this->_oLogger);
+
+        $oShopConfig = Registry::getConfig();
+        $sShopUrl = $oShopConfig->getCurrentShopUrl();
+
+        $sSid = $oSession->sid(true);
+        if ($sSid != '') {
+            $sSid = '&' . $sSid;
+        }
+
+        if (!empty($oShopConfig->getRequestParameter('jsresponse'))) {
```

Comment:
make to methods one to process a "jsResponse" and one for processing "normal responses" then you can create just a isJSResponse and check for it then use one of the new methods to process.

Reply:
please update according to the first review